### PR TITLE
deck: add updateScript

### PR DESCRIPTION
**Pull request by liyangau**
Add  `passthru.updateScript` and _update.sh_ to update short_hash automatically

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by liyangau** [comment]
Hi @Mic92 Thanks for the tip. Would you please review this change?

File: pkgs/applications/networking/deck/default.nix [hunk 1]
```diff
@@ -23,6 +23,8 @@ in buildGoModule rec {
 
   vendorHash = "sha256-jhLZvusYpX5fW1NCmJtwE/p9/wTwzA2hbwt657VsZts=";
 
+  passthru.updateScript = ./update.sh;
```
Comment (NickCao) [inline pkgs/applications/networking/deck/default.nix:26]: Try `nix-update-script`, also we don't really need the `short_hash` thing (it should match the tag?).

Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix:26]: How do you use `nix-update-script`?  I searched around and hardly find anything useful.

I know it is not required, I keep this `short_hash` just to match the official release as possible.

Comment (NickCao) [reply pkgs/applications/networking/deck/default.nix:26]: Just set `passthru.updateScript = nix-update-script { };` and everything magically works.

Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix:26]: > Just set `passthru.updateScript = nix-update-script { };` and everything magically works.
😲 
Just want to clarify, do I still need  `update.sh` in the same folder? If not then how does nix-update knows to pull `short_hash` in the new version?

A little bit of background info.

The reason for using  `passthru.updateScript` is to solve the automate update process does not update `short_hash`. https://github.com/NixOS/nixpkgs/pull/248291

Comment (Mic92) [reply pkgs/applications/networking/deck/default.nix:26]: No it does not do that. You could use nix-update to pull and update the latest version in your script, which saves a few lines of code. But maybe not worth the hassle given that you already wrote the code.

File: pkgs/applications/networking/deck/update.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  echo "Already up to date!"
+  exit 0
+fi
+
+CURRENT_HASH=$(sed -nE 's/\s*short_hash = "(.*)".*/\1/p' ./default.nix)
+NEW_HASH=$(curl -s https://api.github.com/repos/kong/deck/tags | jq -rc '.[] | select(.name=="$DECK_LATEST_VERSION")|.commit.sha' | head -c 7 -)
+
+sed -i "s|$CURRENT_HASH|$NEW_HASH|" "./default.nix"
+
+update-source-version deck "${DECK_LATEST_VERSION//v/}"
```
Comment (Mic92) [inline pkgs/applications/networking/deck/update.sh:20]: Running update for:
 - deck-1.25.0: UPDATING ...
 - deck-1.25.0: ERROR

--- SHOWING ERROR LOG FOR deck-1.25.0 ----------------------

/home/joerg/git/nixpkgs/pkgs/applications/networking/deck/update.sh: line 20: update-source-version: command not found

File: pkgs/applications/networking/deck/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/applications/networking/deck/default.nix:26]: ok. Looks like the update script didn't actually work as of now.

Comment (NickCao) [reply pkgs/applications/networking/deck/default.nix:26]: > The reason for using `passthru.updateScript` is to solve the automate update process does not update `short_hash`. #248291

That's why I said that we don't really need `short_hash` and can just drop it instead.

Comment (Mic92) [reply pkgs/applications/networking/deck/default.nix:26]: I think so too.

File: pkgs/applications/networking/deck/update.sh (hunk 2 above)
Comment (liyangau) [reply pkgs/applications/networking/deck/update.sh:20]: This should be fixed now. 

Can you please tell me how you run the test?

I use `nix-shell maintainers/scripts/update.nix --argstr maintainer liyangau` to test but it did not catch this error.

File: pkgs/applications/networking/deck/default.nix (hunk 1 above)
Comment (liyangau) [reply pkgs/applications/networking/deck/default.nix:26]: I am curious to learn what's the general practice for maintaining a nixpkges for NixOS 

Let's say when there is no compatibility issues, do we do what is minimum to build the package or do we try to build the package as close to official ones as possible?

File: pkgs/applications/networking/deck/update.sh (hunk 2 above)
Comment (Mic92) [reply pkgs/applications/networking/deck/update.sh:20]: I usually do `nix-update --build --commit --use-update-script deck`

Comment (Mic92) [reply pkgs/applications/networking/deck/update.sh:20]: I also downgraded deck beforehand to actually made it upgrade.

Comment (Mic92) [reply pkgs/applications/networking/deck/update.sh:20]: It's still failing:

```
--- SHOWING ERROR LOG FOR deck-1.25.0 ----------------------

this path will be fetched (0.01 MiB download, 0.03 MiB unpacked):
  /nix/store/5w53jaz6408yak154x0qii1xchmknjlv-common-updater-scripts
copying path '/nix/store/5w53jaz6408yak154x0qii1xchmknjlv-common-updater-scripts' from 'https://cache.nixos.org'...
error: cannot evaluate a function that has an argument without a value ('buildGoModule')
       Nix attempted to evaluate a function as a top level expression; in
       this case it must have its arguments supplied either by default
       values, or passed explicitly with '--arg' or '--argstr'. See
       https://nixos.org/manual/nix/stable/language/constructs.html#functions.

       at /home/joerg/git/nixpkgs/pkgs/applications/networking/deck/default.nix:1:3:

            1| { buildGoModule, lib, installShellFiles, fetchFromGitHub }:
             |   ^
            2| let
error: cannot evaluate a function that has an argument without a value ('buildGoModule')
       Nix attempted to evaluate a function as a top level expression; in
       this case it must have its arguments supplied either by default
       values, or passed explicitly with '--arg' or '--argstr'. See
       https://nixos.org/manual/nix/stable/language/constructs.html#functions.

       at /home/joerg/git/nixpkgs/pkgs/applications/networking/deck/default.nix:1:3:

            1| { buildGoModule, lib, installShellFiles, fetchFromGitHub }:
             |   ^
            2| let
error: cannot evaluate a function that has an argument without a value ('buildGoModule')
       Nix attempted to evaluate a function as a top level expression; in
       this case it must have its arguments supplied either by default
       values, or passed explicitly with '--arg' or '--argstr'. See
       https://nixos.org/manual/nix/stable/language/constructs.html#functions.

       at /home/joerg/git/nixpkgs/pkgs/applications/networking/deck/default.nix:1:3:

            1| { buildGoModule, lib, installShellFiles, fetchFromGitHub }:
             |   ^
            2| let
update-source-version: error: Could not find attribute 'deck'!


--- SHOWING ERROR LOG FOR deck-1.25.0 ----------------------
The update script for deck-1.25.0 failed with exit code 1
Traceback (most recent call last):
  File "/home/joerg/git/nix-update/bin/nix-update", line 12, in <module>
    main()
  File "/home/joerg/git/nix-update/nix_update/__init__.py", line 287, in main
    package = update(options)
              ^^^^^^^^^^^^^^^
  File "/home/joerg/git/nix-update/nix_update/update.py", line 321, in update
    run(
  File "/home/joerg/git/nix-update/nix_update/utils.py", line 36, in run
    return subprocess.run(
           ^^^^^^^^^^^^^^^
  File "/nix/store/9hwaihb8rhlvq8s7xq11m204jp51h6xi-python3-3.11.4/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['nix-shell', '/home/joerg/git/nixpkgs/maintainers/scripts/update.nix', '--argstr', 'package', 'deck']' returned non-zero exit status 1.

```

Comment (liyangau) [reply pkgs/applications/networking/deck/update.sh]: Thanks for telling me `nix-update --build --commit --use-update-script deck`. This helps a lot.

The script should work now, can you please have a look. 

Thanks

**Pull request closed without merging**